test/model/src/include/model/kv_transaction_snapshot.h

```diff
@@ -31,45 +31,95 @@
 
 #include <memory>
 #include <unordered_set>
+#include <vector>
 
 #include "model/core.h"
 
 namespace model {
 
+class kv_update;
+
 /*
  * kv_transaction_snapshot --
  *     A transaction snapshot.
  */
 class kv_transaction_snapshot {
 
 public:
+    /*
+     * kv_transaction_snapshot::contains --
+     *     Check whether the given update belongs to the snapshot.
+     */
+    virtual bool contains(const kv_update &update) const noexcept = 0;
+
+protected:
     /*
      * kv_transaction_snapshot::kv_transaction_snapshot --
+     *     Create a new instance of the snapshot.
+     */
+    inline kv_transaction_snapshot(){};
+};
+
+/*
+ * kv_transaction_snapshot_by_exclusion --
+ *     A transaction snapshot based on which transactions to exclude.
+ */
+class kv_transaction_snapshot_by_exclusion : public kv_transaction_snapshot {
+
+public:
+    /*
+     * kv_transaction_snapshot_by_exclusion::kv_transaction_snapshot_by_exclusion --
      *     Create a new instance of the snapshot based on what to exclude.
      */
-    inline kv_transaction_snapshot(
+    inline kv_transaction_snapshot_by_exclusion(
       txn_id_t exclude_after, std::unordered_set<txn_id_t> &&exclude_ids)
         : _exclude_after(exclude_after), _exclude_ids(exclude_ids)
     {
     }
 
     /*
-     * kv_transaction_snapshot::contains --
-     *     Check whether the given transaction ID belongs to the snapshot.
+     * kv_transaction_snapshot_by_exclusion::contains --
+     *     Check whether the given update belongs to the snapshot.
      */
-    inline bool
-    contains(txn_id_t id) const noexcept
-    {
-        if (id > _exclude_after)
-            return false;
-        return _exclude_ids.find(id) == _exclude_ids.end();
-    }
+    virtual bool contains(const kv_update &update) const noexcept override;
 
 private:
     txn_id_t _exclude_after;
     std::unordered_set<txn_id_t> _exclude_ids;
 };
 
+/*
+ * kv_transaction_snapshot_wt --
+ *     A transaction snapshot that emulates WiredTiger's behavior.
+ */
+class kv_transaction_snapshot_wt : public kv_transaction_snapshot {
+
+public:
+    /*
+     * kv_transaction_snapshot_wt::kv_transaction_snapshot_wt --
+     *     Create a new instance of the snapshot.
+     */
+    inline kv_transaction_snapshot_wt(write_gen_t write_gen, txn_id_t snapshot_min,
+      txn_id_t snapshot_max, const std::vector<uint64_t> &snapshots)
+        : _min_id(snapshot_min), _max_id(snapshot_max), _write_gen(write_gen)
+    {
+        for (uint64_t t : snapshots)
+            _include_ids.insert(t);
```

Comment:
Can we use the `std::copy` function here? Or `_include_ids.insert(snapshots.cbegin(), snapshots.cend());`

Reply:
Yes, I just switched to it—I wasn't aware of this function.

For a function that is supposed to save keystrokes, I think it took more characters than what I had there before 🤣